curvefs/src/client/rpcclient/metaserver_client.cpp

```diff
@@ -497,163 +497,185 @@ bool GroupInodeIdByPartition(
+                    CopysetGroupID(poolID, copysetID)));
+            for (const auto &id : it) {
+                request.add_inodeid(id);
+            }
+            if (request.inodeid_size() <= 0) {
+                LOG(WARNING) << "BatchGetInodeXAttr request empty.";
+                return MetaStatusCode::PARAM_ERROR;
+            }
 
-                MetaStatusCode ret = response.statuscode();
-                if (ret != MetaStatusCode::OK) {
-                    LOG_IF(WARNING, ret != MetaStatusCode::NOT_FOUND)
-                        << "BatchGetXAttr failed, errcode = " << ret
-                        << ", errmsg = " << MetaStatusCode_Name(ret);
-                } else if (response.xattr_size() > 0 &&
-                           response.has_appliedindex()) {
-                    auto retXattr = response.xattr();
-                    for_each(retXattr.begin(), retXattr.end(),
-                        [&](XAttr &a) { xattr->push_back(a); });
-                    metaCache_->UpdateApplyIndex(
-                        CopysetGroupID(poolID, copysetID),
-                            response.appliedindex());
-                } else {
-                    LOG(WARNING) << "BatchGetXAttr ok, but"
-                                 << " applyIndex or attr not set in response: "
-                                 << response.DebugString();
-                    return -1;
-                }
-                return ret;
-            };
-            auto taskCtx = std::make_shared<TaskContext>(
-                MetaServerOpType::BatchGetInodeAttr, task, fsId, inodeId);
-            BatchGetInodeAttrExcutor excutor(
-                opt_, metaCache_, channelManager_, taskCtx);
-            auto ret = ConvertToMetaStatusCode(excutor.DoRPCTask());
+            curvefs::metaserver::MetaServerService_Stub stub(channel);
+            stub.BatchGetXAttr(cntl, &request, &response, nullptr);
+
+            if (cntl->Failed()) {
+                metaserverClientMetric_->batchGetXattr.eps.count << 1;
+                LOG(WARNING) << "BatchGetXAttr Failed, errorcode = "
+                                << cntl->ErrorCode()
+                                << ", error content:" << cntl->ErrorText()
+                                << ", log id = " << cntl->log_id();
+                return -cntl->ErrorCode();
+            }
+
+            MetaStatusCode ret = response.statuscode();
             if (ret != MetaStatusCode::OK) {
-                xattr->clear();
-                return ret;
+                LOG_IF(WARNING, ret != MetaStatusCode::NOT_FOUND)
+                    << "BatchGetXAttr failed, errcode = " << ret
+                    << ", errmsg = " << MetaStatusCode_Name(ret);
+            } else if (response.xattr_size() > 0 &&
+                        response.has_appliedindex()) {
+                auto retXattr = response.xattr();
+                for_each(retXattr.begin(), retXattr.end(),
+                    [&](XAttr &a) { xattr->push_back(a); });
```

Comment:
```cpp
auto* attrs = response.mutable_xattr();
xattr->insert(xattr->end(),
              std::make_move_iterator(attrs->begin()),
              std::make_move_iterator(attrs->end()));
```

might be a better way to just move xattrs in response to the destination.

Reply:
fixed